File: entropy.cpp
#include <iostream>
#include <set>
#include <vector>
#include <string>
#include <cmath>

using namespace std;

double entropy(string& st) {
    vector<char> stvec(st.begin(), st.end());
    set<char> alphabet(stvec.begin(), stvec.end());
    vector<double> freqs;
    for (set<char>::iterator c = alphabet.begin(); c != alphabet.end(); ++c) {
        int ctr = 0;
        for (vector<char>::iterator s = stvec.begin(); s != stvec.end(); ++s) {
            if (*s == *c) {
                ++ctr;
            }
        }
        freqs.push_back((double)ctr / (double)stvec.size());
    }
    double ent = 0;
    double ln2 = log(2);
    for (vector<double>::iterator f = freqs.begin(); f != freqs.end(); ++f) {
        ent += *f * log(*f)/ln2;
    }
    ent = -ent;
    return ent;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        cout << "prints the base2 entropy of the argument" << endl;
    }
    string input = string(argv[1]);
    cout << entropy(input) << endl;
}

File: entropy.py
#!/usr/bin/python
#
# Shannon Entropy of a string
# = minimum average number of bits per symbol
# required for encoding the string
#
# So the theoretical limit for data compression:
# Shannon Entropy of the string * string length
# FB - 201011291

import math
import sys

st = sys.argv[1]
#st = 'aabcddddefffg' # input string
# st = '00010101011110' # Shannon entropy for this would be 1 bit/symbol

#print 'Input string:'
#print st
#print
stList = list(st)

alphabet = []
for i in stList:
    if i not in alphabet:
        alphabet.append(i)
#print 'Alphabet of symbols in the string:'
#print(alphabet)
#print
# calculate the frequency of each symbol in the string
freqList = []
for symbol in alphabet:
    ctr = 0
    for sym in stList:
        if sym == symbol:
            ctr += 1
    freqList.append(float(ctr) / len(stList))
#print 'Frequencies of alphabet symbols:'
#print freqList
#print
# Shannon entropy
ent = 0.0
for freq in freqList:
    ent = ent + freq * math.log(freq, 2)
ent = -ent
print(ent)

